libmemcached/version.cc

```diff
@@ -58,182 +58,219 @@ const char * memcached_lib_version(void)
+    if (memcached_failed(rrc))
     {
-      instance->major_version= instance->minor_version= instance->micro_version= UINT8_MAX;
-      rc= MEMCACHED_SOME_ERRORS;
+      errors_happened= true;
+      (void)memcached_set_error(*instance, rrc, MEMCACHED_AT);
       continue;
     }
+    success++;
+  }
 
-    /* Find the space, and then move one past it to copy version */
-    response_ptr= index(buffer, ' ');
-    response_ptr++;
-
-    // UNKNOWN
-    if (*response_ptr == 'U')
+  if (success)
+  {
+    /* Collect the returned items */
+    memcached_server_write_instance_st instance;
+    while ((instance= memcached_io_get_readable_server(ptr)))
     {
-      instance->major_version= instance->minor_version= instance->micro_version= UINT8_MAX;
-      continue;
+      char buffer[32];
+      memcached_return_t rrc= memcached_response(instance, buffer, sizeof(buffer), NULL);
+      if (rrc != MEMCACHED_SUCCESS)
+      {
+        errors_happened= true;
+      }
     }
+  }
+
+  return errors_happened ? MEMCACHED_SOME_ERRORS : MEMCACHED_SUCCESS;
+}
 
-    instance->major_version= (uint8_t)strtol(response_ptr, (char **)NULL, 10);
-    if (errno == ERANGE)
+static inline memcached_return_t version_ascii_instance(memcached_server_write_instance_st instance)
+{
+  memcached_return_t rc= MEMCACHED_SUCCESS;
+  if (instance->major_version == UINT8_MAX)
+  {
+    libmemcached_io_vector_st vector[]=
     {
-      instance->major_version= instance->minor_version= instance->micro_version= UINT8_MAX;
-      rc= MEMCACHED_SOME_ERRORS;
-      continue;
-    }
+      { strlen("version\r\n"), "version\r\n" }
+    };
 
-    response_ptr= index(response_ptr, '.');
-    response_ptr++;
+    uint32_t before_active= instance->cursor_active;
 
-    instance->minor_version= (uint8_t)strtol(response_ptr, (char **)NULL, 10);
-    if (errno == ERANGE)
+    rc= memcached_vdo(instance, vector, 1, true);
+    if (rc == MEMCACHED_WRITE_FAILURE)
     {
-      instance->major_version= instance->minor_version= instance->micro_version= UINT8_MAX;
-      rc= MEMCACHED_SOME_ERRORS;
-      continue;
+      (void)memcached_set_error(*instance, rc, MEMCACHED_AT);
+      memcached_io_reset(instance);
+      return rc;
     }
-
-    response_ptr= index(response_ptr, '.');
-    response_ptr++;
-    instance->micro_version= (uint8_t)strtol(response_ptr, (char **)NULL, 10);
-    if (errno == ERANGE)
+    /* If no_reply or piped is set, cursor_active is not increased in memcached_vdo().
+     * Think calling memcached_version_instance() when connecting to a server
+     * in pipe or no_reply operations.
+     */
+    if (before_active == instance->cursor_active)
     {
-      instance->major_version= instance->minor_version= instance->micro_version= UINT8_MAX;
-      rc= MEMCACHED_SOME_ERRORS;
-      continue;
+      memcached_server_response_increment(instance);
     }
+    char buffer[32];
+    rc= memcached_response(instance, buffer, sizeof(buffer), NULL);
```

Comment:
network 관련 read failure 발생하면, 
memcached_response() 안에서 memcached_io_reset() 시키나요 ?

Reply:
네. 아래 함수들은 memcached_response() 안에서 io read 시에 호출되는 함수들이고, 각 함수 내에서 network failure 발생 시에 memcached_io_reset() (memcached_quit_server()와 동일) 를 수행합니다.
```
 memcached_return_t memcached_io_readline(memcached_server_write_instance_st ptr,
                                          char *buffer_ptr,
                                          size_t size,
                                          size_t& total_nr)
 {
...
       memcached_return_t rc= memcached_io_read(ptr, buffer_ptr, 1, &nread);
       if (memcached_failed(rc) and rc == MEMCACHED_IN_PROGRESS)
       {
         memcached_quit_server(ptr, true);
         return memcached_set_error(*ptr, rc, MEMCACHED_AT);
       }
```
 ```
static memcached_return_t _io_fill(memcached_server_write_instance_st ptr)
 {
   ssize_t data_read;
   do
   {
     data_read= recv(ptr->fd, ptr->read_buffer, MEMCACHED_MAX_BUFFER, MSG_DONTWAIT);
     if (data_read == SOCKET_ERROR)
     {
       switch (get_socket_errno())
       {
...
       default:
         int local_errno= get_socket_errno(); // We cache in case memcached_quit_server() modifies errno
         memcached_quit_server(ptr, true);
         return memcached_set_errno(*ptr, local_errno, MEMCACHED_AT);
       }
     }
     else if (data_read == 0)
     {
...
       memcached_quit_server(ptr, true);
       return memcached_set_error(*ptr, MEMCACHED_CONNECTION_FAILURE, MEMCACHED_AT,
                                  memcached_literal_param("recv() returned zero, server has disconnected"));
     }
   } while (data_read <= 0);
```